File: packages/cli/hieudoanm/swift/Sources/hieudoanm/Commands/Telegram/Telegram.swift
import Foundation
import ArgumentParser

struct Telegram: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "telegram",
        abstract: "Telegram bot tools",
        subcommands: [
            TelegramMessage.self,
            TelegramWebhook.self,
        ]
    )

    mutating func run() {}
}

struct TelegramMessage: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "message",
        abstract: "Send messages",
        subcommands: [TelegramMessageSend.self]
    )

    mutating func run() {}
}

struct TelegramMessageSend: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "send", abstract: "Send a message")

    @Option(name: .long, help: "Bot token")
    var token: String?

    @Option(name: .long, help: "Chat ID")
    var chatID: String?

    @Argument(help: "Message text")
    var text: String

    mutating func run() async throws {
        let token = token ?? prompt("Bot token: ")
        let chatID = chatID ?? prompt("Chat ID: ")

        let bot = TelegramBot(token: token)
        try await bot.sendMessage(chatID: chatID, text: text)
        print("Message sent")
    }
}

struct TelegramWebhook: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "webhook",
        abstract: "Manage webhooks",
        subcommands: [TelegramWebhookSet.self, TelegramWebhookInfo.self, TelegramWebhookDelete.self]
    )

    mutating func run() {}
}

struct TelegramWebhookSet: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "set", abstract: "Set webhook")

    @Option(name: .long, help: "Bot token")
    var token: String?

    @Option(name: .long, help: "Webhook URL")
    var url: String?

    mutating func run() async throws {
        let token = token ?? prompt("Bot token: ")
        let url = url ?? prompt("Webhook URL: ")

        let bot = TelegramBot(token: token)
        try await bot.setWebhook(url: url)
        print("Webhook set")
    }
}

struct TelegramWebhookInfo: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "info", abstract: "Get webhook info")

    @Option(name: .long, help: "Bot token")
    var token: String?

    mutating func run() async throws {
        let token = token ?? prompt("Bot token: ")

        let bot = TelegramBot(token: token)
        let info = try await bot.getWebhookInfo()
        print(info)
    }
}

struct TelegramWebhookDelete: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "delete", abstract: "Delete webhook")

    @Option(name: .long, help: "Bot token")
    var token: String?

    mutating func run() async throws {
        let token = token ?? prompt("Bot token: ")

        let bot = TelegramBot(token: token)
        try await bot.deleteWebhook()
        print("Webhook deleted")
    }
}

private func prompt(_ msg: String) -> String {
    print(msg, terminator: "")
    return readLine() ?? ""
}

File: packages/app/headless/jack/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/Telegram.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.core.Context
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.default
import io.github.hieudoanm.cli.services.Requests

class TelegramCommand : CliktCommand(name = "telegram") {
    override fun help(context: Context) = "Telegram bot and message tools"
    init {
        subcommands(TelegramMessage(), TelegramWebhook())
    }
    override fun run() = Unit
}

class TelegramMessage : CliktCommand(name = "message") {
    override fun help(context: Context) = "Send Telegram messages"
    init {
        subcommands(TelegramMessageSend())
    }
    override fun run() = Unit
}

class TelegramMessageSend : CliktCommand(name = "send") {
    override fun help(context: Context) = "Send a Telegram message"
    override fun run() {
        echo("Telegram Token: ")
        val token = readLine() ?: ""
        echo("Telegram Chat ID: ")
        val chatId = readLine() ?: ""
        echo("Telegram Message: ")
        val message = readLine() ?: ""
        val url = "https://api.telegram.org/bot${token}/sendMessage"
        val body = """{"chat_id":"$chatId","text":"$message"}"""
        val result = Requests.post(url, body)
        result.onSuccess { echo("Success") }
            .onFailure { echo("Error: ${it.message}") }
    }
}

class TelegramWebhook : CliktCommand(name = "webhook") {
    override fun help(context: Context) = "Manage Telegram webhooks"
    init {
        subcommands(TelegramWebhookSet(), TelegramWebhookInfo(), TelegramWebhookDelete())
    }
    override fun run() = Unit
}

class TelegramWebhookSet : CliktCommand(name = "set") {
    override fun help(context: Context) = "Set a Telegram webhook URL"
    override fun run() {
        echo("Telegram Token: ")
        val token = readLine() ?: ""
        echo("Telegram Webhook: ")
        val webhook = readLine() ?: ""
        val url = "https://api.telegram.org/bot${token}/setWebhook"
        val body = """{"url":"$webhook"}"""
        val result = Requests.post(url, body)
        result.onSuccess { echo("Success") }
            .onFailure { echo("Error: ${it.message}") }
    }
}

class TelegramWebhookInfo : CliktCommand(name = "info") {
    override fun help(context: Context) = "Get current webhook info"
    override fun run() {
        echo("Telegram Token: ")
        val token = readLine() ?: ""
        val url = "https://api.telegram.org/bot${token}/getWebhookInfo"
        val result = Requests.post(url, "{}")
        result.onSuccess { echo("Success") }
            .onFailure { echo("Error: ${it.message}") }
    }
}

class TelegramWebhookDelete : CliktCommand(name = "delete") {
    override fun help(context: Context) = "Delete the Telegram webhook"
    override fun run() {
        echo("Telegram Token: ")
        val token = readLine() ?: ""
        val url = "https://api.telegram.org/bot${token}/deleteWebhook"
        val result = Requests.post(url, "{}")
        result.onSuccess { echo("Success") }
            .onFailure { echo("Error: ${it.message}") }
    }
}
